crates/kornia-io/src/tiff.rs

```diff
@@ -0,0 +1,207 @@
+    file_path: impl AsRef<Path>,
+    image: &Image<u8, C>,
+) -> Result<(), IoError>
+where 
+    T: ColorType<Inner = u8>,  // Đảm bảo T::Inner là u8
+    [T::Inner]:TiffValue,   
+{
+    if image.num_channels() != C {
+        return Err(IoError::TiffEncodingError(
+            format!("Image has {} channels, but expected {}", image.num_channels(), C)
+        ));
+
+    }
+    
+    let file = File::create(file_path.as_ref())?;
+    let writer = BufWriter::new(file);
+    let mut tiff = TiffEncoder::new(writer)?;
+    let image_encoder = tiff.new_image::<T>(image.width() as u32, image.height() as u32)?;
+    image_encoder.write_data(image.as_slice())?;
+
+    Ok(())
+
+}
+
+/// Reads a TIFF image with a single channel (rgb8).
+/// #Arguments
+/// file_path - The path to the TIFF file.
+/// 
+pub fn read_image_tiff_rgb8(
+    file_path: impl AsRef<Path>,
+) -> Result<Image<u8, 3>, IoError> {
+    read_image_tiff_internal::<u8, 3>(file_path)
+}
+
+/// Reads a TIFF image with a single channel (gray8).
+/// #Arguments
+/// file_path - The path to the TIFF file.
+pub fn read_image_tiff_gray8(
+    file_path: impl AsRef<Path>,
+) -> Result<Image<u8, 1>, IoError> {
+    read_image_tiff_internal::<u8,1>(file_path)
+}
+
+/// Reads a TIFF image with RGBA8 channels
+/// #Arguments
+/// file_path - The path to the TIFF file.
+pub fn read_image_tiff_rgba8(
+    file_path: impl AsRef<Path>,
+) -> Result<Image<u8, 4>, IoError> {
+    read_image_tiff_internal::<u8,4>(file_path)
+}
+
+/// A trait to convert TIFF values to a specific type.
+/// 
+/// This trait is used to convert TIFF values (u8 or f32) to a specific type (e.g., u8 or f32).
+/// It provides two methods:
+pub trait TiffValueConverter: Sized {
+    /// * `from_u8`: Converts a u8 value to the specific type.
+    fn from_u8(value: u8) -> Self;
+    /// * `from_f32`: Converts a f32 value to the specific type.
+    fn from_f32(value: f32) -> Self;
+}
+
+impl TiffValueConverter for u8 {
+    fn from_u8(value: u8) -> Self { value }
+    fn from_f32(value: f32) -> Self { (value * 255.0) as u8 }
+}
+
+impl TiffValueConverter for f32 {
+    fn from_u8(value: u8) -> Self { (value as f32) / 255.0 }
+    fn from_f32(value: f32) -> Self { value }
+}
+
+// Read a TIFF image with a single channel (rgb8).
+//
+fn read_image_tiff_internal<P:Clone,const N: usize>(
+    file_path: impl AsRef<Path>,
+) -> Result<Image<P,N>, IoError> 
+    where 
+        P: TiffValueConverter,
+{
+    let file_path = file_path.as_ref().to_owned();
+
+    if !file_path.exists() {
+        return Err(IoError::FileDoesNotExist(file_path.to_path_buf()));
+    }
+
+    if file_path.extension().map_or(true, |ext| {
+        !ext.eq_ignore_ascii_case("tif") && !ext.eq_ignore_ascii_case("tiff")
+    }) {
+        return Err(IoError::InvalidFileExtension(file_path.to_path_buf()));
+    }
+
+    let file = File::open(&file_path)?;
+    let mut decoder = Decoder::new(file).map_err(IoError::TiffError)?;
```

Comment:
```suggestion
    let mut decoder = Decoder::new(file)?;
```
forward error diectly like this